File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int lswrc = 0;

        HashMap<Character, Integer> hm = new HashMap<>();
        int left = 0, right = 0;

        while (right < s.length() && left <= right) {
            if (hm.containsKey(s.charAt(right))) {
                lswrc = Math.max(lswrc, right - left);
                int last = hm.get(s.charAt(right));
                hm.replace(s.charAt(right), right);
                left = Math.max(last + 1, left);
                right++;
            } else {
                hm.put(s.charAt(right), right);
                // lswrc = Math.max(lswrc, right - left + 1);
                right++;
            }
        }
        if (right == s.length()) {
            lswrc = Math.max(lswrc, right - left);
        }
        
       
        return lswrc;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        if (n == 0) {
            return 0;
        }
        int ans = INT_MIN;
        map<char, int> mpp;
        int start = 0, end = 0;
        while (end < n) {

            while (end < n && mpp.find(s[end]) == mpp.end()) {
                // cout << bool(mpp.find(s[end]) == mpp.end()) << endl;
                mpp[s[end]] = 1;
                end ++;
                ans = max(ans, end - start);

                cout << start << " " << end << endl;
            }

            while (start <= end && mpp.find(s[end]) != mpp.end()) {
                mpp.erase(s[start]);
                start ++;
                cout << start << " " << end << endl;
            }
        }
        
        
        return ans;

    }
};